src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java

```diff
@@ -0,0 +1,107 @@
+    this.dataReferenceDao = dataReferenceDao;
+    this.samService = samService;
+    this.jobService = jobService;
+    this.validationUtils = validationUtils;
+  }
+
+  public DataReferenceDescription getDataReference(
+      String workspaceId, String referenceId, AuthenticatedUserRequest userReq) {
+
+    authz(userReq, workspaceId, SamUtils.SAM_WORKSPACE_READ_ACTION);
+
+    return dataReferenceDao.getDataReference(UUID.fromString(referenceId));
+  }
+
+  public DataReferenceDescription createDataReference(
+      String workspaceId, CreateDataReferenceRequestBody body, AuthenticatedUserRequest userReq) {
+
+    // validate shape of request as soon as it comes in
+    if ((body.getReferenceType().isPresent() && body.getReference().isPresent())
+        == body.getResourceId().isPresent()) {
+      throw new InvalidDataReferenceException(
+          "Data reference must contain either a resource id or a reference type and a reference description");
+    }
+
+    authz(userReq, workspaceId, SamUtils.SAM_WORKSPACE_WRITE_ACTION);
+
+    UUID referenceId = UUID.randomUUID();
+    String description =
+        "Create data reference " + referenceId.toString() + " in workspace " + workspaceId;
+
+    JobBuilder createJob =
+        jobService
+            .newJob(
+                description,
+                UUID.randomUUID().toString(),
+                CreateDataReferenceFlight.class,
+                body,
+                userReq)
+            .addParameter(DataReferenceFlightMapKeys.REFERENCE_ID, referenceId)
+            .addParameter(DataReferenceFlightMapKeys.WORKSPACE_ID, UUID.fromString(workspaceId))
+            .addParameter(DataReferenceFlightMapKeys.NAME, body.getName())
+            .addParameter(
+                DataReferenceFlightMapKeys.CLONING_INSTRUCTIONS, body.getCloningInstructions());
+
+    if (body.getReferenceType().isPresent() && body.getReference().isPresent()) {
+      String ref =
+          validationUtils.validateReference(
+              DataReferenceDescription.ReferenceTypeEnum.fromValue(body.getReferenceType().get()),
+              body.getReference().get(),
+              userReq);
+
+      createJob.addParameter(
+          DataReferenceFlightMapKeys.REFERENCE_TYPE, body.getReferenceType().get());
+      createJob.addParameter(DataReferenceFlightMapKeys.REFERENCE, ref);
+    }
+
+    if (body.getCredentialId().isPresent()) {
+      createJob.addParameter(
+          DataReferenceFlightMapKeys.CREDENTIAL_ID, body.getCredentialId().get());
+    }
+
+    createJob.submitAndWait(String.class);
+
+    return dataReferenceDao.getDataReference(referenceId);
+  }
+
+  private void authz(AuthenticatedUserRequest userReq, String workspaceId, String action) {
+    try {
+      samService.isAuthorized(
+          userReq.getRequiredToken(), SamUtils.SAM_WORKSPACE_RESOURCE, workspaceId, action);
+    } catch (ApiException samEx) {
+      throw new SamApiException(samEx);
```

Comment:
! This was something I just realized recently -  `isAuthorized` returns authorization status as boolean that's being ignored here. 
The ApiException gets thrown if Sam fails to call the API, but if the user is unauthorized it returns `false` without throwing an error.